libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+void scurves::add_segment_incr_jerk(float tj, float Jp)
+{
+    float Beta = M_PI / tj;
+    float Alpha = Jp / 2.0;
+    float AT = Alpha * tj;
+    float VT = Alpha * (sq(tj) / 2.0 - 2.0 / sq(Beta));
+    float PT = Alpha * ((-1.0 / sq(Beta)) * tj + (1 / 6.0) * powf(tj, 3.0));
+
+    enum jtype_t Jtype = jtype_t::POSITIVE;
+    float J = Jp;
+    float T = segment[num_segs - 1].start_time + tj;
+    float A = segment[num_segs - 1].start_accel + AT;
+    float V = segment[num_segs - 1].start_vel + segment[num_segs - 1].start_accel * tj + VT;
+    float P = segment[num_segs - 1].start_pos + segment[num_segs - 1].start_vel * tj + 0.5 * segment[num_segs - 1].start_accel * sq(tj) + PT;
+    add_segment(T, Jtype, J, A, V, P);
+}
+
+// generate  decreasing jerk magnitude time segment based on a raised cosine profile
+void scurves::add_segment_decr_jerk(float tj, float Jp)
+{
+    float Beta = M_PI / tj;
+    float Alpha = Jp / 2.0;
+    float AT = Alpha * tj;
+    float VT = Alpha * (sq(tj) / 2.0 - 2.0 / sq(Beta));
+    float PT = Alpha * ((-1.0 / sq(Beta)) * tj + (1 / 6.0) * powf(tj, 3.0));
+    float A2T = Jp * tj;
+    float V2T = Jp * sq(tj);
+    float P2T = Alpha * ((-1.0 / sq(Beta)) * 2.0 * tj + (4.0 / 3.0) * powf(tj, 3.0));
+
+    enum jtype_t Jtype = jtype_t::NEGATIVE;
+    float J = Jp;
+    float T = segment[num_segs - 1].start_time + tj;
+    float A = (segment[num_segs - 1].start_accel - AT) + A2T;
+    float V = (segment[num_segs - 1].start_vel - VT) + (segment[num_segs - 1].start_accel - AT) * tj + V2T;
+    float P = (segment[num_segs - 1].start_pos - PT) + (segment[num_segs - 1].start_vel - VT) * tj + 0.5 * (segment[num_segs - 1].start_accel - AT) * sq(tj) + P2T;
+    add_segment(T, Jtype, J, A, V, P);
+}
+
+// generate three time segment raised cosine jerk profile
+void scurves::add_segments_incr_const_decr_jerk(float tj, float Jp, float Tcj)
+{
+    add_segment_incr_jerk(tj, Jp);
+    add_segment_const_jerk(Tcj, Jp);
+    add_segment_decr_jerk(tj, Jp);
+}
+
+// generate time segments for straight segment
+void scurves::add_segments_straight(float Pp)
+{
+    if (is_zero(Pp)) {
+        return;
+    }
+
+    float tj = otj;
+    float Jp = jerk_max;
+    float Ap = accel_max;
+    float Vp = vel_max;
+
+    float t2, t4, t6;
+    cal_posfast(tj, Jp, Ap, Vp, Pp / 2.0, Jp, t2, t4, t6);
+
+    add_segments_incr_const_decr_jerk(tj, Jp, t2);
+    add_segment_const_jerk(t4, 0.0);
+    add_segments_incr_const_decr_jerk(tj, -Jp, t6);
+
+    float t8 = 2.0f * (Pp / 2.0 - segment[num_segs - 1].start_pos) / segment[num_segs - 1].start_vel;
+    add_segment_const_jerk(t8, 0.0);
+
+    add_segments_incr_const_decr_jerk(tj, -Jp, t6);
+    add_segment_const_jerk(t4, 0.0);
+    add_segments_incr_const_decr_jerk(tj, Jp, t2);
+}
+
+// generate time segments to generate large curved corners
+void scurves::add_segments_curved(float Pp, float Pm)
+{
+    if (is_zero(Pp)) {
+        return;
```

Comment:
Should this be an internal error?  Similarly elsewhere.

Reply:
zapped with splines